### DISCUSSION TITLE
#59で作成する構造体や保持場所について

### DISCUSSION THREAD

**Discussion by mcnLeandro** [Recursion-Group-G/Othello#61] [Q&A]
#59 
![sshot](https://user-images.githubusercontent.com/72016706/141790449-cc9a6412-1475-4fd1-927f-25b4bc93d7c7.png)

この前みんなで話した通り、ノードを連結リストで繋いで、そのそれぞれのノードをvalueとして持つハッシュマップを作成して保持しておく必要があるんですが、みんなの意見が聞きたいです！

- 連結リストはどこの保持するか(おそらくHeadのみ)
- ハッシュマップはどこに保持するか
- 連結リストとハッシュマップを操作する構造体を作成することになるが、その構造体は名前空間であるべきかそれともインスタンス化するクラスか
- 必要なメソッドはあるか(関数の分解でできるであろう細かいメソッドは除く)

これらについて意見が欲しいです！

自分的には構造体が名前空間として状態を維持するのがいいのかなと考えていました！

```js
class ?? {
     static head : SquareNode
     static hashmap: [key: Square, value SquareNode]
     
    //ハッシュマップのkeyはSquare自体にしていたんですけどオブジェクトってkeyとして区別されるんでしたっけ？ 
     static removeSquareNode(square: Square): void
     static addSquareNode(square:Square):void
     static getSquareNode(square:Square): SquareNode
}

```

みんなはどう思いますか？？？？

**Comment by m2mt-l** [comment]
自分も名前空間でよいような気もします。
もしhashmapがここ以外で使うのであればインスタンス化してもよい気がしたのですが、そんな場面ありますっけ？
あと空いている場所を探すメソッドはSquareNodeの方に乗せたほうがよいでしょうか？

**Reply by mcnLeandro** [reply]
空いてる場所を探すメソッドは二つ用意したくて、
一つは
- 特定のSquareNodeから8方向を探す
二つ目
- おいてあるStone全てをもとに全検索
二つ目の意図としては、ゲーム開始の時に最初に発火させたり、
localStorageから黄色い部分を取得できない場合にロードし直したときにこういうメソッド必要かなと考えていました。

空いてるメソッドを探す場所はどちらでもいい気がしてくるんですが、そこでSquareNodeがModelなのかModuleなのかできまる気がしますね。Modelであれば。メソッドを持つことはできないのでSquareNodeを管理している構造体に保つことになりそうです。。

**Comment by m2mt-l** [comment]
Player毎にこの構造体を持つ必要はないですよね？もし持つのであれば以下のようにしてもよい気がしました。
https://recursionist.io/dashboard/course/4/lesson/579

**Reply by mcnLeandro** [reply]
そうです！プレイヤー毎に持つ必要はありません！
この構造体はゲーム全体で共有している画像でいう黄色い部分。「全体から見て。石を置ける可能性のある場所」を管理する構造体になります！

**Reply by m2mt-l** [reply]
プレイヤー毎に石を置ける可能性のある場所が変わるので、プレイヤー毎に持ってもよいかなと思いましたが、どちらにせよターン毎に変わるのでメモリの無駄使いですね笑

**Reply by mcnLeandro** [reply]
あ、そうです！
それに関してはあってもいいかなと思います！Playerごとに置ける場所が違うので！
最初はfor文で毎回黄色い部分を検索する形になると思うんですが、うまくキャッシュできたらそのキャッシュはプレイヤーが持つことになりますね！

ただ、そのタスクは一歩先になりそうです！キャッシュの話も一旦後でAIのキャッシュと一緒に話したほうが良さそうです！
https://whimsical.com/6YrTUaVgeaZS8tgFPNu8jL

**Comment by kuro1221** [comment]
- 連結リストはheadのみを保存。
- メソッドは今の所それ以上思いつかないです

 ちなみにオブジェクトのkeyはStringかrSymbolみたいですね。
https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Map#objects_vs._maps

ローカルストレージから取得することを考えると、構造体は名前空間よりもインスタンス化した方が取扱いやすいのかなと思いました。

**Reply by mcnLeandro** [reply]
javascriptにもメモリアドレス参照があれば。。。
対応策はちょっとめんどくさいですけど、Squareに一意のkeyを持たせることですかね？？つまり idですね for分で作成するときに指定できそうです。もしくはstatic メンバ変数を+1しながら指定できそうですね。

インスタンス化については否定できないですね！
Recursionではインスタンス化していたので！

今回、アプリケーションで絶対一つしかないBoardもインスタンス化しているので統一したほうがいいかもしれないっていうのもありますね。

**Reply by kuro1221** [reply]
Squareに一意のkeyを持たせる必要がありそうですね。
個人的には座標を元にidを作成したいです。
例）(0,0）なら00、(3,4)なら34

**Reply by mcnLeandro** [reply]
あ、それでもいけそうですね！

**Reply by m2mt-l** [reply]
ここの話を聞いていたらインスタンス化した方がよい気もしてきました。
Squareに一意のkeyを持たせる場合以下が必要になりますでしょうか？
・Squareにid: numberのメンバー変数追加
・idを作成するメソッド

idを作成するメソッドはPointがあるのでSquare内で実施してもよい気がしました(this.id= createId()のような)。

**Reply by mcnLeandro** [reply]
そうですね！
コンストラクタで(x,y)を受け取るときにそれをもとに作ってもいいですし、メソッドにしてもらっても問題ないです！

**Comment by m2mt-l** [comment]
enclosure自体を作成するメソッド(createEnclosure)と連結リストを作成するメソッド(linkEnclosure)はここ(EnclosureController)に作ったほうがいいでしょうか？
それとも別でEnclosureBuilderみたいなmoduleを作ってそちらで作成した方がよいでしょうか？

**Reply by mcnLeandro** [reply]
それに関してはEnclosureController内で問題ありません！ 👍

**Comment by m2mt-l** [comment]
isEmptyとは別にもう一つフラグを作成する話はこの構造体の話をしていた時にしましたっけ？
初期状態のEnclosureを最初の石から8方向で確認して作ることを考えているのですが、isEmptyだけだと既に作られたものの重複が判別出来ないと思いました。。

**Reply by m2mt-l** [reply]
もう一つの案としては座標(Squareのid)を取得して、左上->右上->右下->左下->左上のようにほんとうに四角で囲って作成する方法。これだと座標間違ってた時に少し怖いのですが。。

**Reply by mcnLeandro** [reply]
これはenclosureの連結内にSquareが存在するかということですか？？
最初は「isPutableみたいな変数がいる」という話になったんですが、
黄色い部分として登録されているかどうかは
enclosureController内のハッシュマップとSuqareの持っているキーを使って存在チェックすれば確認できるので問題ないです！

**Reply by mcnLeandro** [reply]
個人的には初期状態のEnclosureはboard内を全検索しちゃっていい気がしました。。
メソッドの再利用性がなくならない程度には最適化したいですが、思いつかないうちは一旦全検索でもいいのかと。。。

> 左上->右上->右下->左下->左上のようにほんとうに四角で囲って作成する方法。これだと座標間違ってた時に少し怖いのですが。。

はっきりイメージできていないですが、
Squareは8方向に連結してるので、座標の問題はあまりない気がしました！

**Reply by mcnLeandro** [reply]
```js
class EnclosureController {
  this.hashmap
  this.head
  this.tail
  
  updateEnclosureFromSquare(suquare: Square){
    //Square自身と連結している8方向の追加をする
    
    // Suquareを受け取って必要に応じてEnclosureとして連結する匿名関数作成
   
    //あとはSquare自身とそのSquareの8方向をその匿名関数に通す
  }
  
}
```
自分のイメージではこんな関数↑を作成して
その中で（匿名関数内で）
```js
if(this.hashmap[suquare.id] === null)return;
```
こんな感じで判定できると思ってます！

これを使ってさらに初期化するメソッドとかも作れると思います！

**Reply by m2mt-l** [reply]
石を8方向にハッシュマップから調べる方が他で応用ができるのでそちらにしました！

**Comment by m2mt-l** [comment]
置いた場所から石を返すメソッドは別モジュールの方がよいでしょうか？
